Hackage Revision Request: base-compat-batteries-0.12.2 to allow OneTuple < 0.5

This would help us out in nixpkgs where we now follow Stackage LTS 21 for the versions of base-compat-batteries and OneTuple to use which are 0.12.* and 0.4.*, respectively. For GHCs < 9.0 OneTuple is required to build base-compat-batteries and we run into this problem. Using the two versions together seems to work, so the bound on hackage could be loosened.

Hm. I'm not sure if simply lifting the `OneTuple` upper version bounds in `base-compat-batteries-0.12.2` would be wise, since that could lead to a conditional API. This is because in `OneTuple-0.3.*` and earlier, `Solo`'s data constructor is named `Solo`, but in `OneTuple-0.4.*`, `Solo`'s data constructor is named `MkSolo`. This can be seen in this program:

```hs
#!/usr/bin/env cabal
{- cabal:
build-depends: base, base-compat-batteries==0.12.2, OneTuple==0.3.*
-}
{- project:
with-compiler: ghc-8.10
-}
{-# LANGUAGE PackageImports #-}
module Main (main) where

import "base-compat-batteries" Data.Tuple.Compat (Solo(..))

f :: a -> Solo a
f = MkSolo

main :: IO ()
main = print (f ())
```

Note that I have intentionally set up the project information to ensure that when this program imports `Data.Tuple.Compat`, it will use the `Solo` that is re-exported by `OneTuple-0.3.*`. (This is why I use GHC 8.10, as newer GHCs would bypass `OneTuple`.) This program will fail to compile, since `OneTuple-0.3.*` doesn't define `MkSolo`:

```
$ ./Foo1.hs

C:\msys64\home\winferno\Foo1.hs:14:5: error:
    * Data constructor not in scope: MkSolo :: a -> Solo a
    * Perhaps you meant `Solo' (imported from Data.Tuple.Compat)
   |
14 | f = MkSolo
   |     ^^^^^^
```

On the other hand, if I tweak the program such that `base-compat-batteries-0.12.2` is allowed to build with `OneTuple-0.4.*`:

```hs
#!/usr/bin/env cabal
{- cabal:
build-depends: base, base-compat-batteries==0.12.2, OneTuple==0.4.*
-}
{- project:
with-compiler: ghc-8.10
allow-newer: base-compat-batteries:OneTuple
-}
-- same as before...
```

Then the program _will_ compile, with no changes made to the code!

```
$ ./Foo2.hs
MkSolo ()
```

This could lead to confusing scenarios. On the other hand, the situation is arguably _already_ confusing, since you would get the same thing if you switched from GHC 8.10 to GHC 9.6:

```hs
#!/usr/bin/env cabal
{- cabal:
build-depends: base, base-compat-batteries==0.12.2
-}
{- project:
with-compiler: ghc-9.6
-}
-- same as before
```
```
$ ./Foo3.hs
MkSolo ()
```

--- a/base-compat-batteries/CHANGES.markdown
+++ b/base-compat-batteries/CHANGES.markdown
@@ -1,3 +1,6 @@
+## Changes in 0.12.3 [????.??.??]
+ - Allow building with `OneTuple-0.4.*`.
+
 ## Changes in 0.12.2 [2022.08.11]
  - This coincides with the `base-compat-0.12.2` release. Refer to the
    [`base-compat` changelog](https://github.com/haskell-compat/base-compat/blob/master/base-compat/CHANGES.markdown#changes-in-0122-20220811)

--- a/base-compat-batteries/base-compat-batteries.cabal
+++ b/base-compat-batteries/base-compat-batteries.cabal
@@ -1,5 +1,5 @@
 name:             base-compat-batteries
-version:          0.12.2
+version:          0.12.3
 license:          MIT
 license-file:     LICENSE
 copyright:        (c) 2012-2018 Simon Hengel,
@@ -64,7 +64,7 @@ library
       Haskell2010
   build-depends:
       base        >= 4.3 && < 5,
-      base-compat == 0.12.2,
+      base-compat == 0.12.3,
       ghc-prim
   if !impl(ghc >= 7.8)
     build-depends:
@@ -88,7 +88,17 @@ library
       contravariant >= 1.5 && < 1.6
   if !impl(ghc >= 9.0)
     build-depends:
-      OneTuple >= 0.3 && < 0.4
+      OneTuple >= 0.3 && < 0.5
+
+    -- We want to consistently use `Solo` as the name of the data constructor
+    -- for the base-compat-0.12.* series when building with a pre-9.6 version
+    -- of GHC. This is only possible on pre-7.8 versions of GHC (which do not
+    -- support pattern synonyms) with OneTuple-0.3.*, which defines a `Solo`
+    -- data constructor. (In OneTuple-0.4.* and later, the data constructor is
+    -- instead named `MkSolo`.)
+    if !impl(ghc >= 7.8)
+      build-depends:
+        OneTuple == 0.3.*
   ghc-options:
       -fno-warn-duplicate-exports
   if impl(ghc >= 7.10)
@@ -256,6 +266,7 @@ test-suite spec
       -- Other tests
       SafeHaskellSpec
       TestHspecTrustworthy
+      T91Spec
   build-depends:
       base >= 4.3 && < 5
     , base-compat-batteries

--- a/base-compat-batteries/src/Data/Tuple/Compat.hs
+++ b/base-compat-batteries/src/Data/Tuple/Compat.hs
@@ -1,6 +1,23 @@
 {-# LANGUAGE CPP, NoImplicitPrelude, PackageImports #-}
+#if __GLASGOW_HASKELL__ >= 708
+{-# LANGUAGE PatternSynonyms #-}
+#endif
 module Data.Tuple.Compat
-  ( Solo (..)
+  (
+#if MIN_VERSION_ghc_prim(0,10,0)
+    Solo(MkSolo, Solo)
+#elif __GLASGOW_HASKELL__ >= 708 && \
+      __GLASGOW_HASKELL__ < 800 && \
+      defined(MIN_VERSION_OneTuple)
+# if MIN_VERSION_OneTuple(0,4,0)
+    Solo
+  , pattern Solo
+# else
+    Solo(Solo)
+# endif
+#else
+    Solo(Solo)
+#endif
   , fst
   , snd
   , curry
@@ -12,5 +29,5 @@ module Data.Tuple.Compat
 import "base-compat" Data.Tuple.Compat
 #else
 import "base" Data.Tuple
-import "OneTuple" Data.Tuple.Solo (Solo(..))
+import "OneTuple" Data.Tuple.Solo
 #endif

--- a/base-compat-batteries/test/T91Spec.hs
+++ b/base-compat-batteries/test/T91Spec.hs
@@ -1,0 +1,15 @@
+{-# OPTIONS_GHC -fno-warn-deprecations #-}
+-- | A regression test for https://github.com/haskell-compat/base-compat/issues/91.
+module T91Spec (main, spec) where
+
+import Data.Tuple.Compat
+import Test.Hspec (Expectation, Spec, describe, hspec, it)
+
+main :: IO ()
+main = hspec spec
+
+spec :: Spec
+spec =
+  describe "Solo data constructor" $
+    it "is always available" $ do
+      Solo () `seq` return () :: Expectation

--- a/base-compat/CHANGES.markdown
+++ b/base-compat/CHANGES.markdown
@@ -1,3 +1,8 @@
+## Changes in 0.12.3 [????.??.??]
+ - This coincides with the `base-compat-batteries-0.12.3` release. Refer to the
+   [`base-compat-batteries` changelog](https://github.com/haskell-compat/base-compat/blob/master/base-compat-batteries/CHANGES.markdown#changes-in-0123-????????)
+   for more details.
+
 ## Changes in 0.12.2 [2022.08.11]
  - Sync with `base-4.17`/GHC 9.4
  - Backport `(.^.)`, `(.>>.)`, `(.<<.)`, `(!>>.)`, `(!<<.)`, `oneBits` to

--- a/base-compat/base-compat.cabal
+++ b/base-compat/base-compat.cabal
@@ -1,5 +1,5 @@
 name:             base-compat
-version:          0.12.2
+version:          0.12.3
 license:          MIT
 license-file:     LICENSE
 copyright:        (c) 2012-2018 Simon Hengel,

--- a/base-compat/src/Data/Tuple/Compat.hs
+++ b/base-compat/src/Data/Tuple/Compat.hs
@@ -8,8 +8,10 @@ module Data.Tuple.Compat
   , curry
   , uncurry
   , swap
-#if MIN_VERSION_ghc_prim(0,7,0)
-  , Solo(..)
+#if MIN_VERSION_ghc_prim(0,10,0)
+  , Solo(MkSolo, Solo)
+#elif MIN_VERSION_ghc_prim(0,7,0)
+  , Solo(Solo)
 #endif
   ) where
 

